app/extensions/restManager/RestManager.py

```diff
@@ -0,0 +1,367 @@
+    ENDPOINTS = None
+    NAME = None
+
+    def __init__(self, app, pre_initialize=False, *args, **kwargs):
+        super(RestManager, self).__init__(*args, **kwargs)
+        self.initialized = False
+
+        # Must be overwritten by the derived class
+        assert self.NAME is not None
+        assert self.ENDPOINT_PREFIX is not None
+        assert self.ENDPOINTS is not None
+
+        self.app = app
+        self.targets = set([])
+
+        app.edm = self
+
+        if pre_initialize:
+            self._ensure_initialized()
+
+    def _endpoint_fmtstr(self, tag, target='default'):
+        endpoint_tag_fmtstr = self._endpoint_tag_fmtstr(tag)
+        assert endpoint_tag_fmtstr is not None, 'The endpoint tag was not recognized'
+
+        if endpoint_tag_fmtstr.startswith('//'):
+            endpoint_tag_fmtstr = endpoint_tag_fmtstr[2:]
+            endpoint_tag_fmtstr = f'{self.ENDPOINT_PREFIX}/{endpoint_tag_fmtstr}'
+
+        endpoint_url_ = self.get_target_endpoint_url(target)
+        endpoint_fmtstr = f'{endpoint_url_}/{endpoint_tag_fmtstr}'
+        return endpoint_fmtstr
+
+    def _endpoint_tag_fmtstr(self, tag):
+        endpoint = self.ENDPOINTS
+
+        component_list = tag.split('.')
+        for component in component_list:
+            try:
+                endpoint_ = endpoint.get(component, None)
+            except Exception:
+                endpoint_ = None
+
+            if endpoint_ is None:
+                break
+
+            endpoint = endpoint_
+
+        return endpoint
+
+    def _parse_config_uris(self):
+        """Obtain and verify the required configuration settings and
+        update the list of known EDM targets by URI.
+
+        This procedure is primarily focused on matching a URI to
+        authentication credentials. It ignores matching in the other direction.
+
+        """
+        uris = self.app.config.get(f'{self.NAME}_URIS', {})
+        authns = self.app.config.get(f'{self.NAME}_AUTHENTICATIONS', {})
+
+        # Check for the 'default' EDM
+        assert uris.get('default'), f"Missing a 'default' {self.NAME}_URI"
+        has_required_default_authn = (
+            isinstance(authns.get('default'), dict)
+            and authns['default'].get('username')
+            and authns['default'].get('password')
+        )
+        assertion_msg = f"Missing {self.NAME}_AUTHENTICATION credentials for 'default'"
+        assert has_required_default_authn, assertion_msg
+
+        # Check URIs have matching credentials
+        missing_creds = [k for k in uris.keys() if not authns.get(k)]
+        assert (
+            not missing_creds
+        ), f"Missing credentials for named {self.NAME} configs: {', '.join(missing_creds)}"
+
+        # Update the list of known named targets
+        self.targets.update(uris.keys())
+
+        # Assign local references to the configuration settings
+        self.uris = uris
+        self.auths = authns
+
+    def _init_sessions(self):
```

Comment:
Not certain yet if this needs to be in the base or derived class and if the 'session.login' needs to be configurable. Currently not working for ACM.